controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1424,6 +1437,199 @@ func (r *BareMetalHostReconciler) manageHostPower(prov provisioner.Provisioner,
 }
 
+// DataImage handler for attaching/detaching image.
+func (r *BareMetalHostReconciler) handleDataImageActions(prov provisioner.Provisioner, info *reconcileInfo) actionResult {
+	dataImage := &metal3api.DataImage{}
+	if err := r.Get(info.ctx, info.request.NamespacedName, dataImage); err != nil {
+		// DataImage does not exist or it may have been deleted
+		if k8serrors.IsNotFound(err) {
+			info.log.Info("dataImage not found")
+			return nil
+		}
+		// Error reading the object - requeue the request.
+		return actionError{fmt.Errorf("could not load dataImage, %w", err)}
+	}
+
+	// Set ControllerReference to DataImage
+	if !ownerReferenceExists(info.host, dataImage) {
+		if err := controllerutil.SetControllerReference(info.host, dataImage, r.Scheme()); err != nil {
+			return actionError{fmt.Errorf("could not set bmh as controller, %w", err)}
+		}
+		if err := r.Update(info.ctx, dataImage); err != nil {
+			return actionError{fmt.Errorf("failure creating dataImage resource, %w", err)}
+		}
+		// TODO(hroyrh) : Should we requeue at this point
+		// return actionContinue{}
+		return actionContinue{dataImageUpdateDelay}
+	}
+
+	// Initialize empty dataImage status
+	if dataImage.Status.AttachedImage == nil {
+		dataImage.Status.AttachedImage = &metal3api.AttachedImageReference{}
+	}
+	if dataImage.Status.Error == nil {
+		dataImage.Status.Error = &metal3api.DataImageError{}
+	}
+
+	// Check if any attach/detach action is pending or failed to attach
+	nodeReservation, nodeLastError := prov.GetDataImageStatus()
+	if nodeReservation != "" {
+		info.log.Info("Node is already under reservation, requeue", "Reserved by", nodeReservation)
+		return actionContinue{dataImageRetryDelay}
+	}
+
+	// Is the current dataImage status valid
+	dirty := false
+	// In case the last node error was not nil for dataimage,
+	// upadate message and counter
+	if nodeLastError != "" {
+		dataImage.Status.Error.Message = nodeLastError
+		dataImage.Status.Error.Count++
+		dirty = true
+	}
+
+	deleteDataImage := false
+	if !dataImage.DeletionTimestamp.IsZero() {
+		deleteDataImage = true
+	}
+
+	requestedURL := dataImage.Spec.URL
+
+	// We can assume non null value since we initialized the status earlier
+	attachedURL := dataImage.Status.AttachedImage.URL
+
+	if deleteDataImage {
+		info.log.Info("DataImage requested for deletion")
+		if attachedURL != "" || dirty {
+			info.log.Info("Detaching DataImage as it was deleted")
+			err := r.detachDataImage(prov, info, dataImage)
+			if err != nil {
+				return actionError{fmt.Errorf("failed to detach, %w", err)}
+			}
+
+			// Requeue to give time to the DataImage Reconciler to update the
+			// status. In case of failure, we will enter this section and
+			// detachDataImage will be called again -> can this cause issues ?
+			return actionContinue{dataImageUpdateDelay}
+		}
+
+		// In case there was no DataImage attached we simply exit
+		// Should we use actionContinue or simply return nil ?
```

Comment:
This is a case for `nil` since nothing has been done here and no more actions are needed.

Reply:
Makes sense.